nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/admin/StartServerHook.java

```diff
@@ -0,0 +1,266 @@
+        GlassFishLogManager logManager = GlassFishLogManager.getLogManager();
+        if (logManager == null) {
+            return;
+        }
+        logManager.getAllHandlers().forEach(Handler::close);
+    }
+
+
+    private void startGlassFishInstance() {
+        try {
+            this.builder.start();
+        } catch (Exception e) {
+            throw new IllegalStateException("GlassFish instance startup failed.", e);
+        }
+    }
+
+
+    private void waitForThreads(Predicate<Thread> filter) {
+        final Thread[] threads = getThreadsToDie(filter);
+        log(() -> "Waiting for shutdown of threads:\n" + toString(threads));
+        while(true) {
+            Thread thread = getFirstAlive(threads);
+            if (thread == null) {
+                break;
+            }
+            try {
+                thread.join();
+                log(() -> "Joined thread " + toString(thread));
+            } catch (InterruptedException e) {
+                log(e);
+                continue;
+            }
+        }
+    }
+
+
+    private Thread[] getThreadsToDie(Predicate<Thread> filter) {
+        final Thread[] threads = new Thread[Thread.activeCount() + 10];
+        Thread.enumerate(threads);
+        log(() -> "Found threads:\n" + toString(threads));
+        return Stream.of(threads).filter(Objects::nonNull).filter(t -> t != this).filter(t -> !t.isDaemon())
+            .filter(filter).toArray(Thread[]::new);
+    }
+
+
+    private Thread getFirstAlive(Thread[] shutdown) {
+        for (Thread thread : shutdown) {
+            if (thread.isAlive()) {
+                return thread;
+            }
+        }
+        return null;
+    }
+
+
+    private void log(Supplier<String> message) {
+        if (logFile == null) {
+            return;
+        }
+        logFile.append(Instant.now().toString()).append(' ').append(message.get()).append('\n');
+        logFile.flush();
+    }
+
+
+    private void log(Exception e) {
+        if (logFile == null) {
+            return;
+        }
+        logFile.append(Instant.now().toString()).append(' ').append(e.getMessage()).append('\n');
+        e.printStackTrace(logFile);
+        logFile.append('\n').flush();
+    }
+
+
+    private static PrintStream getLogFileOld(Instant startTime) {
+        if (!LOG_RESTART) {
+            return null;
+        }
+        try {
+            return new PrintStream(LOGDIR.resolve("restart-" + startTime + "-old.log").toFile());
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+
+    private static ProcessBuilder prepareStartup(Instant startTime, String classpath, String[] sysprops, String classname,
+        String[] args) {
+        final Path javaExecutable = detectJavaExecutable();
+        final List<String> cmdline = new ArrayList<>();
+        if (!OS.isWindows()) {
+            cmdline.add("nohup");
+        }
+        cmdline.add(javaExecutable.toFile().getAbsolutePath());
```

Comment:
Why not use just `Path.toAbsolutePath()` or `Path.toAbsolutePath().normalize()`?

Reply:
It doesn't have same result on all Windows if I remember well. Path always uses /, on current Windows "it depends".